modules/packages/LinearAlgebra.chpl

```diff
@@ -605,6 +604,96 @@ private inline proc _raw(D: domain(1), i) {
 // Matrix Structure
 //
 
+
+/*
+   Return a Vector containing the diagonal elements of ``A`` if the argument ``A`` is of rank 2.
+   Return a diagonal Matrix whose diagonal contains elements of ``A`` if argument ``A`` is of rank 1.
+ */
+proc diag(A: [?Adom] ?eltType, k=0){
+  //This should be printed at compile time-"This function only supports 0-based non-strided arrays, 
+  //including the vectors and matrices created from this module."
+
+  if(Adom.rank == 2) then{
+    if(k==0) then{
+      return _diag_vec(A);  
+    }
+    else{
+      return _diag_vec(A, k);
+    }
+  }
+  else if(Adom.rank == 1) then {
+    return _diag_mat(A);
+  }
+  else{
+    compilerError("A must have rank 2 or less");
+  }
+
+}
+
+proc _diag_vec(A:[?Adom] ?eltType){
+
+  //better way to do this if type of Adom.dim(1) and Adom.dim(2) can be determined
+  if(Adom.dim(1).length<Adom.dim(2).length) then{
+    var diagonal = Vector(Adom.dim(1).length, eltType);
+    forall i in Adom.dim(1) do{
+      diagonal[i] = A[i,i];
+    }
+    return diagonal;
+  }
+  else{
+    var diagonal = Vector(Adom.dim(2).length, eltType);
+    forall i in Adom.dim(2) do{
+      diagonal[i] = A[i,i];
+    }
+    return diagonal;
+  }
+}
+
+proc _diag_vec(A:[?Adom] ?eltType, k){
```

Comment:
I would add `private` to these `_diag_*` functions. Though, we can do this as a short follow-up PR.